pkg/kubevirt/core/core.go

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package kubevirt contains the cloud kubevirt specific implementations to manage machines
+package core
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	api "github.com/gardener/machine-controller-manager-provider-kubevirt/pkg/kubevirt/apis"
+	clouderrors "github.com/gardener/machine-controller-manager-provider-kubevirt/pkg/kubevirt/errors"
+	"github.com/gardener/machine-controller-manager-provider-kubevirt/pkg/kubevirt/util"
+
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+	utilpointer "k8s.io/utils/pointer"
+	kubevirtv1 "kubevirt.io/client-go/api/v1"
+	cdi "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const ProviderName = "kubevirt"
+
+// PluginSPIImpl is the real implementation of PluginSPI interface
+// that makes the calls to the provider SDK
+type PluginSPIImpl struct {
+	client client.Client
+}
+
+// NewPluginSPIImpl creates a new kubevirt cloud provider based on the passed client or secret.
+func NewPluginSPIImpl(client client.Client, secret *corev1.Secret) (*PluginSPIImpl, error) {
```

Comment:
I don't see this method called anywhere except in tests? How is the client initialized in the production case? I don't think it is.
Also, I am not particularly fond of passing both a client and a secret, especially since you can create the client from the secret. I would remove the second parameter and make the client creation external to the constructor, or I would add a second constructor with a secret and call the first one from it.

Reply:
Yeah I still need to do some refactoring. The problem is, we followed the vSphere implementation, where in the previous one, we were creating the client from the secret which is passed to all methods that the constructor has. Now, this is not possible because the secret is not passed at the time where the client is created. And the client is created from the secret because it has the kubeconfig for the kiubevirt cluster!!